rust/src/dcerpc/dcerpc.rs

```diff
@@ -728,33 +876,67 @@ impl DCERPCState {
         } else {
             self.query_completed = true;
         }
-        // TODO buffer extra data when input length is more than fraglen
         parsed = self.bytes_consumed as i32;
 
+        let current_call_id = self.get_hdr_call_id().unwrap_or(0);
+        let endianness = self.get_hdr_drep_0() & 0x10;
+
         match self.get_hdr_type() {
             Some(x) => match x {
                 DCERPC_TYPE_BIND | DCERPC_TYPE_ALTER_CONTEXT => {
                     retval = self.process_bind_pdu(&buffer[parsed as usize..]);
                     if retval == -1 {
                         return AppLayerResult::err();
                     }
+                    self.update_tx_id(current_call_id);
                 }
                 DCERPC_TYPE_BINDACK | DCERPC_TYPE_ALTER_CONTEXT_RESP => {
+                    let hdrtype = self.get_hdr_type().unwrap_or(DCERPC_TYPE_BINDACK);
+                    let create_tx = self.find_tx(current_call_id);
+                    match create_tx {
+                        Some(mut transaction) => {
+                            transaction.resp_cmd = hdrtype;
+                        }
+                        None => {
+                            let mut tx =
+                                DCERPCTransaction::new(self.tx_id, current_call_id, endianness);
+                            tx.resp_cmd = hdrtype;
+                            self.transactions.push(tx);
+                        }
+                    };
                     retval = self.process_bindack_pdu(&buffer[parsed as usize..]);
                     if retval == -1 {
                         return AppLayerResult::err();
                     }
+                    if let Some(ref mut tx) = self.find_tx(current_call_id) {
```

Comment:
Naming scheme: Yes, it is. Possible result of code updation and not variable names.
It is because in case `create_tx` was `None` i.e. no tx with the required `call_id` exists then we create a new transaction and add it to the state. Since it is a newly added object in the state, it'll have to be looked up again.
I tried using a common variable to store either of them but could not make Rust happy about the diff transaction types required (`DCERPCTransaction`) while the creation and while using an existing one (`&mut DCERPCTransaction`)
I cannot think of a nicer way to do this. Do you have anything in mind?

Reply:
I think this will be functionally the same, except there is no double search
```diff
diff --git a/rust/src/dcerpc/dcerpc.rs b/rust/src/dcerpc/dcerpc.rs
index dc1ccac53..0442e44b6 100644
--- a/rust/src/dcerpc/dcerpc.rs
+++ b/rust/src/dcerpc/dcerpc.rs
@@ -891,27 +891,22 @@ impl DCERPCState {
                     self.update_tx_id(current_call_id);
                 }
                 DCERPC_TYPE_BINDACK | DCERPC_TYPE_ALTER_CONTEXT_RESP => {
+                    retval = self.process_bindack_pdu(&buffer[parsed as usize..]);
                     let hdrtype = self.get_hdr_type().unwrap_or(DCERPC_TYPE_BINDACK);
-                    let create_tx = self.find_tx(current_call_id);
-                    match create_tx {
-                        Some(mut transaction) => {
-                            transaction.resp_cmd = hdrtype;
-                        }
-                        None => {
-                            let mut tx =
-                                DCERPCTransaction::new(self.tx_id, current_call_id, endianness);
-                            tx.resp_cmd = hdrtype;
-                            self.transactions.push(tx);
-                        }
+                    let tx = if let Some(mut tx) = self.find_tx(current_call_id) {
+                        tx.resp_cmd = hdrtype;
+                        tx
+                    } else {
+                        let mut tx = DCERPCTransaction::new(self.tx_id, current_call_id, endianness);
+                        tx.resp_cmd = hdrtype;
+                        self.transactions.push(tx);
+                        self.transactions.last_mut().unwrap()
                     };
-                    retval = self.process_bindack_pdu(&buffer[parsed as usize..]);
                     if retval == -1 {
                         return AppLayerResult::err();
                     }
-                    if let Some(ref mut tx) = self.find_tx(current_call_id) {
-                        tx.resp_done = true;
-                        tx.frag_cnt_tc += 1;
-                    }
+                    tx.resp_done = true;
+                    tx.frag_cnt_tc += 1;
                     self.update_tx_id(current_call_id);
                 }
                 DCERPC_TYPE_REQUEST => {
```
By moving the call to `self.process_bindack_pdu` to the top the borrow checker accepts it. The same logic fails to compile with keeping `self.process_bindack_pdu` in the middle. But as it looks like there are no dependencies between them, other than the result check, I think it should be ok.

Reply:
Thank you very much!